service-base/src/main/java/org/eclipse/hono/service/amqp/RequestResponseEndpoint.java

```diff
@@ -162,34 +179,95 @@ protected final void handleMessage(final ProtonConnection con, final ProtonRecei
         } else {
             formalCheck.fail(new AmqpErrorException(AmqpError.DECODE_ERROR, "malformed payload"));
         }
+
         final HonoUser clientPrincipal = Constants.getClientPrincipal(con);
-        formalCheck.compose(ok -> isAuthorized(clientPrincipal, targetAddress, message)).compose(authorized -> {
-
-            logger.debug("client [{}] is {}authorized to {}:{}", clientPrincipal.getName(), authorized ? "" : "not ",
-                    targetAddress, message.getSubject());
-            if (authorized) {
-                try {
-                    processRequest(message, targetAddress, clientPrincipal);
-                    ProtonHelper.accepted(delivery, true);
-                    return Future.succeededFuture();
-                } catch (final DecodeException e) {
-                    return Future.failedFuture(new AmqpErrorException(AmqpError.DECODE_ERROR, "malformed payload"));
-                }
-            } else {
-                return Future.failedFuture(new AmqpErrorException(AmqpError.UNAUTHORIZED_ACCESS, "unauthorized"));
-            }
-
-        }).otherwise(t -> {
-
-            if (t instanceof AmqpErrorException) {
-                final AmqpErrorException cause = (AmqpErrorException) t;
-                MessageHelper.rejected(delivery, cause.asErrorCondition());
-            } else {
-                logger.debug("error processing request [resource: {}, op: {}]: {}", targetAddress, message.getSubject(), t.getMessage());
-                MessageHelper.rejected(delivery, ProtonHelper.condition(AmqpError.INTERNAL_ERROR, "internal error"));
-            }
-            return null;
-        });
+
+        final String replyTo = message.getReplyTo();
+        allocateReceiverForReplyTo(replyTo, receiver);
```

Comment:
In theory, a client should always use the same reply-to address for messages sent over this link. However, nothing prevents the client from opening multiple receiver links and sending messages using arbitrary reply-to addresses that map to its multiple receiver links. We should either try to prevent that (how?) or we should also keep track of the reply-to addresses per receiver so that we can remove the entries from the reply-to/receiver map more efficiently when a receiver is closed/disconnected, i.e. without having to iterate over all of the map's entries.
WDYT?

Reply:
If a client would use multiple reply-to addresses, then right now, I think it should still work fine. As each receiver will replenish credits properly once the request has been processed.

Of course it might get interesting when the client starts re-using the same reply-to address with several connections. In this case it wouldn't be possible anymore look up the correct receiver instance to replenish.

I am not so sure another map will help much here. You would need to re-lookup all elements again in the reply-to/receiver map. On the other hand, you could simply iterate over the entrySet of the map and kick out receivers by their connection:

~~~java
private void deallocateAllReceiversForConnection(final ProtonConnection connection) {
    replyToReceiverMap
            .entrySet()
            .removeIf(entry -> entry.getValue().getSession().getConnection() == connection);
}
~~~

Reply:
On theory client could also re-use the same reply-to address. That would probably cause much more problems, as they could steal credits from each other.

The problem is, you could only really mitigate this by doing a check, if the reply-to address doesn't have any existing receiver set with a different connection. Otherwise rejecting the request. But that would make each request a bit more expensive.

Reply:
> The problem is, you could only really mitigate this by doing a check, if the reply-to address doesn't have any existing receiver set with a different connection. Otherwise rejecting the request. But that would make each request a bit more expensive.

This would only need to be checked when the client opens its receiver link for the reply-to address, so it would need to be done only seldomly ....

Reply:
> I am not so sure another map will help much here. You would need to re-lookup all elements again in the reply-to/receiver map. On the other hand, you could simply iterate over the entrySet of the map and kick out receivers by their connection

You are probably right, since we are not talking about too many clients anyways. the clients are basically protocol adapter instances, so we are probably talking dozens to several hundred entries only...

Reply:
> > The problem is, you could only really mitigate this by doing a check, if the reply-to address doesn't have any existing receiver set with a different connection. Otherwise rejecting the request. But that would make each request a bit more expensive.
>
> This would only need to be checked when the client opens its receiver link for the reply-to address, so it would need to be done only seldomly ....

I don't think that is correct, or maybe I didn't understand what you had in mind.

When you have multiple receivers (on the endpoint side), receiving messages with the same reply-to address, then only one of the receivers will receive new credits, and too many. All others will be stuck. And this can only be checked when the incoming request is received.

Reply:
Ok ... I understand ... well, that assumes that the client wouldn't send any request if the open of its receiver link failed. However, we are currently talking about a misbehaving client. So IMHO the problem should be handled at the earliest possible point, when the request is received. At that point it should be checked if the request may be processed, or rejected.

Reply:
then we also need to keep track of the client's receiver(s) and the source address they have been opened for. Then you could check for every incoming message whether there is a corresponding receiver link ...

Reply:
I am not sure I understand what you have in mind. Assuming the following scenario:

* Client A opens receiver to "reply-to-A"
* Client A opens sender to "request"
* Client B open sender to "request" with reply-to="reply-to-A"
* Client B sends request

How would that help?

Reply:
I implemented that change anyway. However I am not sure that it helps much in the described scenario.

Reply:
It helps with the following scenario 

>  Ok ... I understand ... well, that assumes that the client wouldn't send any request if the open of its receiver link failed. However, we are currently talking about a misbehaving client. So IMHO the problem should be handled at the earliest possible point, when the request is received. At that point it should be checked if the request may be processed, or rejected.

1. Client A opens receiver for *reply-to-A*
2. Client A opens sender with target address *registration/my-tenant*
3. Client A sends message with reply-to address *no-receiver-available*

Reply:
> But I don't see how to fix the other issue

what is *the other issue*?

Reply:
> what is the other issue?

* Client A opens receiver to "reply-to-A"
* Client A opens sender to "request"
* Client B open senders to "request"
* Client B sends request with reply-to="reply-to-A"

Reply:
IMHO we could address this by checking if the corresponding links belong to the same session.

Reply:
I am not sure if I would do that. If we decide to put the service behind a qpid router, then that would no longer work, would it?

Reply:
maybe not

Can we try to focus on the original problem that we wanted to solve. We currently do not check if the links belongs together as well, so I would rather have a solution for the problem at hand in the 0.7 release and deal with other improvements later, ok?

Reply:
Yes, I agree, we should clean that up at a later time.

So the PR should be fine in general. I think all the code style, documentation issues should be addresses as well. The only thing left is fixing the unit tests, or any issues which they bring up.